### PR TITLE
Refactor Harvester to the background task

### PR DESCRIPTION
**Pull request by amdomanska**
https://github.com/DOAJ/doajPM/issues/2153

READY FOR INITIAL REVIEW.

Still to be done: 
- [x] Unit Tests
- [x] Change of app.logger to job.audit_logs

### CODE REVIEW

File: portality/tasks/harvester.py
Comment (richard-jones) [inline portality/tasks/harvester.py:95]: I don't think this is needed any more

File: portality/tasks/harvester.py
Comment (richard-jones) [inline portality/tasks/harvester.py:116]: suggest that we put this on the `long_running` queue instead, as the harvester can run for a while

File: portality/tasks/harvester.py
Comment (richard-jones) [inline portality/tasks/harvester.py:124]: `long_running` queue here too

**Review by richard-jones — CHANGES_REQUESTED** [review]

File: doajtest/unit/apiv2_tests/test_apiv2_discovery.py
Comment (richard-jones) [inline doajtest/unit/apiv2_tests/test_apiv2_discovery.py:1]: erroneous character?

File: doajtest/unit/resources/harvester_resp.json
Comment (richard-jones) [inline doajtest/unit/resources/harvester_resp.json:1]: Very unimportant, but it might be best to pretty print this so we can easily inspect it by eye during development down the line

File: doajtest/unit/test_harvester.py
Comment (richard-jones) [inline doajtest/unit/test_harvester.py:28]: I think we decided that this was no longer necessary, and we should handle it by injecting the correct data into the database instead

File: doajtest/unit/test_harvester.py
Comment (richard-jones) [inline doajtest/unit/test_harvester.py:50]: Now that we figured out why the blocking save wasn't working, I'd go back to using that in the journal save

File: doajtest/unit/test_harvester.py
Comment (richard-jones) [inline doajtest/unit/test_harvester.py:58]: the httputil.get doesn't seem to get overridden by the mock

File: doajtest/unit/test_harvester.py
Comment (richard-jones) [inline doajtest/unit/test_harvester.py:77]: I think we planned to only override httputil.get, so that our test also works out the `query` method

File: portality/tasks/harvester_helpers/workflow.py
Comment (richard-jones) [inline portality/tasks/harvester_helpers/workflow.py:32]: An observation from another thread of work - we should probably only consider journals that are `in_doaj`, so you can add `in_doaj=False` to this method call, I think.

File: portality/tasks/harvester_helpers/workflow.py
Comment (richard-jones) [inline portality/tasks/harvester_helpers/workflow.py:44]: I'm not sure I understand what's happening here.  The HarvestState object should have a method/property `suspended`.  Having something that's for test only that seems to work around the HarvestState object API seems odd and risky.  Instead if there's a problem with what the object looks like in the test, then we should fix the test.

**Review by richard-jones — CHANGES_REQUESTED** [review]

File: doajtest/unit/test_harvester.py
Comment (amdomanska) [reply doajtest/unit/test_harvester.py:77]: Yes, true. But I haven't found a way to get to work correctly with Mocks. I will write a separate test to test query method only to get that cover (that is a separate issue, that I feel like my test covers a way too big chunk of code, probably something to think about later).

File: portality/tasks/harvester_helpers/workflow.py
Comment (amdomanska) [reply portality/tasks/harvester_helpers/workflow.py:32]: ok, but I will add that as the default parameter so it could be called differently if needed.

File: portality/tasks/harvester_helpers/workflow.py
Comment (amdomanska) [reply portality/tasks/harvester_helpers/workflow.py:44]: very true. I meant to change that and forgot about that part, sorry!

File: doajtest/unit/test_harvester.py
Comment (amdomanska) [reply doajtest/unit/test_harvester.py:28]: Yes, it is unnecessary

File: portality/tasks/harvester_helpers/workflow.py
Comment (amdomanska) [reply portality/tasks/harvester_helpers/workflow.py:32]: The `Journal.issns_by_owner` method does not accept `in_doaj` param:
```
@classmethod
    def issns_by_owner(cls, owner):
        q = IssnQuery(owner)
        res = cls.query(q=q.query())
        issns = [term.get("term") for term in res.get("facets", {}).get("issns", {}).get("terms", [])]
        return issns
```
I think it is actually a good idea to have it. I'll add the implementation.

File: portality/models/background.py
Comment (Steven-Eardley) [inline portality/models/background.py:87]: do we want print statements rather than logging?

File: portality/settings.py
Comment (Steven-Eardley) [reply portality/settings.py:1031]: bump for this, we can't merge as-is.

File: portality/tasks/harvester_helpers/workflow.py
Comment (Steven-Eardley) [inline portality/tasks/harvester_helpers/workflow.py:62]: We won't have the sigterm control flow any more

**Review by Steven-Eardley — CHANGES_REQUESTED** [review]
Main thing is removing the process name and sigterm control flow. They were to ensure we don't get more than one harvester running at the same time. I suppose it could still happen with background jobs, but none of our other tasks check whether there's a previous run in progress.

File: portality/models/background.py
Comment (amdomanska) [reply portality/models/background.py:87]: no, we're logging them as we used to and I just added the print statement at the end of log to make test debugging easier.

File: portality/tasks/harvester_helpers/epmc/client.py
Comment (amdomanska) [inline portality/tasks/harvester_helpers/epmc/client.py:130]: This diff is very weirdly presented. All that happened here is the extraction of url_from_query method outside of query method.

File: .gitignore
Comment (richard-jones) [inline .gitignore:119]: Should this file be in the gitignore?  It's also in the PR

File: portality/models/background.py
Comment (richard-jones) [inline portality/models/background.py:87]: Just a reminder note to remove print statements

File: portality/models/harvester.py
Comment (richard-jones) [inline portality/models/harvester.py:11]: job has been added to the signature, but it doesn't seem to be used anywhere inside the model

File: portality/models/v2/journal.py
Comment (richard-jones) [inline portality/models/v2/journal.py:975]: If this is set, it should be a boolean `True` not string `"True"`

File: portality/models/v2/journal.py
Comment (richard-jones) [inline portality/models/v2/journal.py:1003]: Same here, should just be the boolean

File: portality/models/v2/journal.py
Comment (richard-jones) [inline portality/models/v2/journal.py:1000]: Does adding this default to `False` change any other usages of this query?  The previous version didn't consider `in_doaj` at all, so returned everything.  Now we will definitely only return in_doaj=True or in_doaj=False values.

Also, if the default value is valid at all, the default to `False` seems risky.  You should probably have to intentionally ask for things that are not in DOAJ, otherwise you should just get the stuff that is.

File: portality/models/v2/journal.py
Comment (richard-jones) [inline portality/models/v2/journal.py:73]: The default to `None` here is at odds with the default to `False` of its usage with the `ISSNQuery`

File: portality/tasks/consumer_main_queue.py
Comment (richard-jones) [inline portality/tasks/consumer_main_queue.py:27]: As we're putting these on the long_running queue, they should be imported in `consumer_long_running.py` instead

File: portality/tasks/harvester_helpers/epmc/client.py
Comment (richard-jones) [inline portality/tasks/harvester_helpers/epmc/client.py:61]: The EMPC client should not know about our job queue.  For example, what if we want to run the client independently of the job queue, or integrate it into some other part of the system as well?  Architecturally this is not a good approach, it creates a two-way binding between modules which should only have a one-way binding.

The client should be a generic client, and the job queue should wrangle it.  Looking into why the job is passed in, it's for the recording of logs, and for no other functional purpose.  If that's the case I'd suggest 2 alternative approaches:

1. pass in a "logger" for the client to use.  This could be provided by the job queue to route the logs to the job's audit trail
2. Have the client log to a state variable (e.g. EuropePMC.log) and have the task read the logs into its own audit logs after the event.

File: portality/tasks/harvester_helpers/epmc/epmc_harvester.py
Comment (richard-jones) [inline portality/tasks/harvester_helpers/epmc/epmc_harvester.py:14]: See comments in previous file - job should not be passed in here

File: portality/tasks/harvester_helpers/workflow.py
Comment (richard-jones) [inline portality/tasks/harvester_helpers/workflow.py:16]: Another place where job should not be passed in

**Review by richard-jones — CHANGES_REQUESTED** [review]
A few odds and ends, and one main architectural issue that we should address before merge

File: .gitignore
Comment (amdomanska) [reply .gitignore:119]: Well, this file is changed every single time the harvester test is run and I don't think we want to have it in every single PR later. That is why I added that to the repo and then set github to ignore it. Not sure what is the best solution here though.

File: portality/models/background.py
Comment (amdomanska) [reply portality/models/background.py:87]: I actually added the print message to make debugging of the background jobs easier. I will add `if app.config.get("DOAJENV") == 'dev'` to make sure it prints only in dev.

File: portality/tasks/harvester_helpers/epmc/client.py
Comment (richard-jones) [inline portality/tasks/harvester_helpers/epmc/client.py:71]: Probably no need to issue the warning here, just write to the `app.logger` if no custom logger is supplied

File: portality/models/background.py
Comment (richard-jones) [reply portality/models/background.py:88]: probably use `if app.config.get("DEBUG") == True` for more generality.

File: portality/tasks/harvester_helpers/workflow.py
Comment (richard-jones) [inline portality/tasks/harvester_helpers/workflow.py:22]: For simplicity, I'd consider just keeping the log information in memory, as it's going to get written to the audit logs down the line anyway.  That saves you just writing it to disk.

If you are certain you want to write it to disk, then we should use our file store instead.  There is a Temporary File store implementation which can be used to stage files.

```
from portality import store
import io

tmpStore = store.StoreFactory.tmp()
tmpStore.store("harvester", "Harvester_logger-2021-06-15.log", source_stream=io.StringIO(""))
loggerFile = tmpStore.get("harvester", "Harvester_logger-2021-06-15.log")

loggerFile.write("log line")
...

# then later
tmpStore.delete_file("harvester", "Harvester_logger-2021-06-15.log")

File: portality/tasks/harvester_helpers/workflow.py
Comment (richard-jones) [inline portality/tasks/harvester_helpers/workflow.py:13]: It doesn't feel right to create a per-run logger on a global singleton member variable (it may have side-effects which could cause runtime errors down the line).  We should either:

1. Pass the logger in wherever it is needed by a `@classmethod`
2. Change the HarvesterWorkflow class so that it is instantiated when it is run, with a logger as an instance member variable

**Review by richard-jones — CHANGES_REQUESTED** [review]
Added a few comments on the implementation of the logging

**Comment by amdomanska** [comment]
@richard-jones another run on logger implementation

**Review by richard-jones — APPROVED** [review]
Reviewed this with @amdomanska yesterday, changes look good to go

File: portality/tasks/harvester_helpers/runner.py
Comment (Steven-Eardley) [inline portality/tasks/harvester_helpers/runner.py:86]: An observation here - we still use `HARVESTER_API_KEYS` which is a dictionary of `{user_id: doaj_api_key}`, but we're no longer creating items via the API. the configuration structure no longer makes sense. (no problem they're just ignored) We may wish to change this to contain which harvester(s) to run per account.

We might want to consider adding a role 'harvester' so it's easier to see and manage which accounts retrieve via the harvester.

**Comment by Steven-Eardley** [comment]
testing successful on the test server (using manual run)

**Pull request merged**
